# Types (estructuras *tipo*)

**NOTA:** Este notebook, y seguramente los que seguirán, serán escritos usando Julia v0.4-dev o v0.4-pre; algunas de los resultados que se obtienen abajo dependen de la versión que se utilice.

En Julia, **todo** es un objeto de algún *tipo*. Por ejemplo:

In [ ]:
typeof( 7 )

In [ ]:
typeof( 18612381238101715161911126301710 )

In [ ]:
typeof( 186123812381017151619111263.01710 )

In [15]:
typeof( 3//1 )

Rational{Int64}

In [16]:
typeof( BigInt(3)//1 )

Rational{Base.GMP.BigInt}

In [ ]:
typeof( π )

In [ ]:
typeof( [1,2,3] )

In [ ]:
typeof( (2,3.2) )

In [ ]:
typeof( ["perro",2,3] )

In [14]:
typeof( [[0, 1]  [1, 0]])

Array{Int64,2}

In [ ]:
typeof( "abc" )

In [ ]:
typeof( "abc"[1] )

In [ ]:
typeof("π")

In [ ]:
f(x) = x^2-3
typeof(f)

In [ ]:
typeof(3.2 + 4im)

In [ ]:
typeof(3//1 + 1im//2)

En Julia, realmente **todo** es un objeto de algún *tipo*.

In [ ]:
typeof(+)

In [ ]:
typeof(//)

La última intrucción *implica* que `+` es alguna función, y como tal, debe funcionar con ciertos tipos de entradas; a esto se le distingue como *métodos*. Para saber qué métodos están definidos en una función, usamos la *función* `methods`:

In [ ]:
typeof(1 == 0)

In [61]:
methods(+)

# 149 methods for generic function "+":
+(x::Bool) at bool.jl:33
+(x::Bool, y::Bool) at bool.jl:36
+(y::AbstractFloat, x::Bool) at bool.jl:46
+(x::Int64, y::Int64) at int.jl:8
+(x::Int8, y::Int8) at int.jl:16
+(x::UInt8, y::UInt8) at int.jl:16
+(x::Int16, y::Int16) at int.jl:16
+(x::UInt16, y::UInt16) at int.jl:16
+(x::Int32, y::Int32) at int.jl:16
+(x::UInt32, y::UInt32) at int.jl:16
+(x::UInt64, y::UInt64) at int.jl:16
+(x::Int128, y::Int128) at int.jl:16
+(x::UInt128, y::UInt128) at int.jl:16
+(x::Integer, y::Ptr{T}) at pointer.jl:77
+(x::Float32, y::Float32) at float.jl:203
+(x::Float64, y::Float64) at float.jl:204
+(z::Complex{T<:Real}, w::Complex{T<:Real}) at complex.jl:112
+(x::Real, z::Complex{T<:Real}) at complex.jl:122
+(z::Complex{T<:Real}, x::Real) at complex.jl:123
+(x::Rational{T<:Integer}, y::Rational{T<:Integer}) at rational.jl:170
+(x::Bool, A::AbstractArray{Bool,N}) at arraymath.jl:136
+(x::Integer, y::Char) at char.jl:44
+(a::Float16, b::Float16) at float16.jl:136
+(x::Base.GMP.BigInt, y::Base.GMP.BigInt) at gmp.jl:246
+(a::Base.GMP.BigInt, b::Base.GMP.BigInt, c::Base.GMP.BigInt) at gmp.jl:269
+(a::Base.GMP.BigInt, b::Base.GMP.BigInt, c::Base.GMP.BigInt, d::Base.GMP.BigInt) at gmp.jl:275
+(a::Base.GMP.BigInt, b::Base.GMP.BigInt, c::Base.GMP.BigInt, d::Base.GMP.BigInt, e::Base.GMP.BigInt) at gmp.jl:282
+(x::Base.GMP.BigInt, c::Union{UInt16,UInt32,UInt64,UInt8}) at gmp.jl:294
+(c::Union{UInt16,UInt32,UInt64,UInt8}, x::Base.GMP.BigInt) at gmp.jl:298
+(x::Base.GMP.BigInt, c::Union{Int16,Int32,Int64,Int8}) at gmp.jl:310
+(c::Union{Int16,Int32,Int64,Int8}, x::Base.GMP.BigInt) at gmp.jl:311
+(x::Base.MPFR.BigFloat, y::Base.MPFR.BigFloat) at mpfr.jl:208
+(x::Base.MPFR.BigFloat, c::Union{UInt16,UInt32,UInt64,UInt8}) at mpfr.jl:215
+(c::Union{UInt16,UInt32,UInt64,UInt8}, x::Base.MPFR.BigFloat) at mpfr.jl:219
+(x::Base.MPFR.BigFloat, c::Union{Int16,Int32,Int64,Int8}) at mpfr.jl:223
+(c::Union{Int16,Int32,Int64,Int8}, x::Base.MPFR.BigFloat) at mpfr.jl:227
+(x::Base.MPFR.BigFloat, c::Union{Float16,Float32,Float64}) at mpfr.jl:231
+(c::Union{Float16,Float32,Float64}, x::Base.MPFR.BigFloat) at mpfr.jl:235
+(x::Base.MPFR.BigFloat, c::Base.GMP.BigInt) at mpfr.jl:239
+(c::Base.GMP.BigInt, x::Base.MPFR.BigFloat) at mpfr.jl:243
+(a::Base.MPFR.BigFloat, b::Base.MPFR.BigFloat, c::Base.MPFR.BigFloat) at mpfr.jl:379
+(a::Base.MPFR.BigFloat, b::Base.MPFR.BigFloat, c::Base.MPFR.BigFloat, d::Base.MPFR.BigFloat) at mpfr.jl:385
+(a::Base.MPFR.BigFloat, b::Base.MPFR.BigFloat, c::Base.MPFR.BigFloat, d::Base.MPFR.BigFloat, e::Base.MPFR.BigFloat) at mpfr.jl:392
+(x::Irrational{sym}, y::Irrational{sym}) at irrationals.jl:72
+(x::Number) at operators.jl:73
+{T<:Number}(x::T<:Number, y::T<:Number) at promotion.jl:211
+{T<:AbstractFloat}(x::Bool, y::T<:AbstractFloat) at bool.jl:43
+(x::Number, y::Number) at promotion.jl:167
+(r1::OrdinalRange{T,S}, r2::OrdinalRange{T,S}) at operators.jl:324
+{T<:AbstractFloat}(r1::FloatRange{T<:AbstractFloat}, r2::FloatRange{T<:AbstractFloat}) at operators.jl:331
+{T<:AbstractFloat}(r1::LinSpace{T<:AbstractFloat}, r2::LinSpace{T<:AbstractFloat}) at operators.jl:350
+(r1::Union{FloatRange{T<:AbstractFloat},LinSpace{T<:AbstractFloat},OrdinalRange{T,S}}, r2::Union{FloatRange{T<:AbstractFloat},LinSpace{T<:AbstractFloat},OrdinalRange{T,S}}) at operators.jl:363
+(x::Ptr{T}, y::Integer) at pointer.jl:75
+{S,T}(A::Range{S}, B::Range{T}) at arraymath.jl:69
+{S,T}(A::Range{S}, B::AbstractArray{T,N}) at arraymath.jl:87
+(A::BitArray{N}, B::BitArray{N}) at bitarray.jl:845
+{T}(B::BitArray{2}, J::Base.LinAlg.UniformScaling{T}) at linalg/uniformscaling.jl:28
+(A::Array{T,2}, B::Base.LinAlg.Diagonal{T}) at linalg/special.jl:122
+(A::Array{T,2}, B::Base.LinAlg.Bidiagonal{T}) at linalg/special.jl:122
+(A::Array{T,2}, B::Base.LinAlg.Tridiagonal{T}) at linalg/special.jl:122
+(A::Array{T,2}, B::Base.LinAlg.SymTridiagonal{T}) at linalg/special.jl:131
+(A::Array{T,N}, B::Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer}) at sparse/s

El resultado de la última celda muestra que hay varios métodos (149) definidos en este momento en la sesión que estoy usando. Además, muestra un montón de otros tipos que no se habían mencionado.

El punto es que hay tipos concretos, como `Float64` o `Int32`, y *también* tipos abstractos, como `Number` o `Integer`.

Los tipos se organizan en un árbol jerárquico. Todo es de tipo `Any`, y la ausencia de 
tipo es `Void` (en Julia v0.3, `Void` es `Nothing`).

In [ ]:
isa(3.2,Any)

In [ ]:
isa(3, Float64)

In [ ]:
isa(nothing, Void)

In [ ]:
nn() = return

typeof( nn() )

Para saber la estructura del árbol jerárquico, uno utiliza `super` para saber lo que está arriba, o `subtypes`

In [ ]:
super(Float64)

In [ ]:
super(AbstractFloat)

(En Julia v0.3, el resultado anterior es `FloatingPoint`.)

In [ ]:
super(AbstractFloat)

In [ ]:
super(Real)

In [ ]:
super(Number)

Y ahora los subtipos:

In [ ]:
subtypes(Integer)

En Julia v0.3 el resultado de la celda anterior es
```julia
5-element Array{Any,1}:
 BigInt  
 Bool    
 Char    
 Signed  
 Unsigned
```

In [ ]:
subtypes(Bool)

In [11]:
subtypes(Real)

4-element Array{Any,1}:
 AbstractFloat       
 Integer             
 Irrational{sym}     
 Rational{T<:Integer}

In [ ]:
subtypes(Number)

In [ ]:
complex(1,4)

In [ ]:
typeof(ans)

# Construyendo tipos propios (*tipos compuestos*)

En Julia, al igual que en todo lenguaje moderno que se precie de serlo, uno puede definir sus propias estructuras, lo que también se llama *tipos compuestos*. Esto da un manejo más versatil de los datos que definen las cantidades en que estamos interesados.

Para definir una nueva estructura de tipo usamos `type`.

Para ejemplificar esto, construiremos primero un vector de dos dimensiones, que llamaremos `Vec2Dtemp`

In [1]:
type Vec2Dtemp
    x
    y
end

In [3]:
vv = Vec2Dtemp(1,2)

Vec2Dtemp(1,2)

In [4]:
typeof(vv)

Vec2Dtemp

In [5]:
vv.x, vv.y

(1,2)

La sutileza, que de hecho es un *problema*, de la definición anterior es que es *demasiado general*. Por ejemplo:

In [6]:
Vec2Dtemp("perro", 1)

Vec2Dtemp("perro",1)

In [7]:
ans.x

"perro"

Quizás esto es lo que queremos, pero en general queremos darle cierta estructura definida a nuestros datos, cosa que en Julia significa ayudarle al compilador a hacer código eficiente. En este caso particular, nos interesa lidiar con números únicamente.

In [8]:
type Vec2DTT
    x :: Real
    y :: Real
end

In [12]:
Vec2DTT(1, 2.3)

Vec2DTT(1,2.3)

In [13]:
typeof(ans.x), typeof(ans.y)

(Int64,Float64)

In [17]:
Vec2DTT("gato", 2.3)

LoadError: LoadError: MethodError: `convert` has no method matching convert(::Type{Real}, ::ASCIIString)
This may have arisen from a call to the constructor Real(...),
since type constructors fall back to convert methods.
Closest candidates are:
  call{T}(::Type{T}, ::Any)
  convert{T<:Real}(::Type{T<:Real}, !Matched::Complex{T<:Real})
  convert{T<:Number}(::Type{T<:Number}, !Matched::Char)
  ...
while loading In[17], in expression starting on line 1

In [18]:
Int64 <: Real

true

In [19]:
typeof(<:)

Function

In [23]:
type Vec2d4
    x :: Int
    y :: Float64
end

In [27]:
Vec2d4(1.1,2)

LoadError: LoadError: InexactError()
while loading In[27], in expression starting on line 1

In [28]:
typeof(Float64)

DataType

In [29]:
pr(x) = (println(x); x)

pr (generic function with 1 method)

In [31]:
pr(2 == 4)

false


false

In [33]:
pr(2>3) && pr(3>2)

false


false

In [34]:
type Vec2D{ T <: Real }
    x :: T
    y :: T
end

In [53]:
aa = Vec2D(1.2,2.3)

Vec2D{Float64}(1.2,2.3)

In [54]:
typeof(aa)

Vec2D{Float64}

In [36]:
Vec2D(-6, 5)

Vec2D{Int64}(-6,5)

Vale la pena notar que `Vec2D` distingue si sus elementos son `Int64` o `Float64`, u otra cosa..

Pero hay diferencias importantes:

In [37]:
Vec2D("perro", 2)

LoadError: LoadError: MethodError: `convert` has no method matching convert(::Type{Vec2D{T<:Real}}, ::ASCIIString, ::Int64)
This may have arisen from a call to the constructor Vec2D{T<:Real}(...),
since type constructors fall back to convert methods.
Closest candidates are:
  Vec2D{T<:Real}(!Matched::T<:Real, ::T<:Real)
  call{T}(::Type{T}, ::Any)
  convert{T}(::Type{T}, !Matched::T)
while loading In[37], in expression starting on line 1

In [40]:
Vec2D(1, 2.0)

LoadError: LoadError: MethodError: `convert` has no method matching convert(::Type{Vec2D{T<:Real}}, ::Int64, ::Float64)
This may have arisen from a call to the constructor Vec2D{T<:Real}(...),
since type constructors fall back to convert methods.
Closest candidates are:
  Vec2D{T<:Real}(::T<:Real, !Matched::T<:Real)
  call{T}(::Type{T}, ::Any)
  convert{T}(::Type{T}, !Matched::T)
while loading In[40], in expression starting on line 1

Para resolver el último caso, usamos `promote`

In [55]:
promote(1.2, 2)

(1.2,2.0)

In [50]:
Vec2D( promote(1.2,2) )

LoadError: LoadError: MethodError: `convert` has no method matching convert(::Type{Vec2D{T<:Real}}, ::Tuple{Float64,Float64})
This may have arisen from a call to the constructor Vec2D{T<:Real}(...),
since type constructors fall back to convert methods.
Closest candidates are:
  call{T}(::Type{T}, ::Any)
  convert{T}(::Type{T}, !Matched::T)
  Vec2D{T<:Real}(!Matched::T<:Real, !Matched::T<:Real)
while loading In[50], in expression starting on line 1

Aquí nos ayuda el operador *splat*, `...`, que son precisamente los (tres) puntos suspensivos:

In [52]:
Vec2D( promote(1.2,2)... )

Vec2D{Float64}(1.2,2.0)

Para evitar hacer esto a cada rato, definimos un nuevo método para `Vec2D`

In [56]:
methodswith(Vec2D)

0-element Array{Method,1}

In [57]:
Vec2D(a,b) = Vec2D(promote(a,b)...)

Vec2D{T<:Real}

In [59]:
methods(Vec2D)

4-element Array{Any,1}:
 call{T<:Real}(::Type{Vec2D{T<:Real}}, x::T<:Real, y::T<:Real) at In[34]:2
 call(::Type{Vec2D{T<:Real}}, a, b) at In[57]:1                           
 call{T}(::Type{T}, arg) at essentials.jl:57                              
 call{T}(::Type{T}, args...) at essentials.jl:58                          

In [60]:
Vec2D(1, 2.3)

Vec2D{Float64}(1.0,2.3)

In [62]:
promote(1im, 2.3)

(0.0 + 1.0im,2.3 + 0.0im)

In [64]:
Vec2D(1, 2.3)

Vec2D{Float64}(1.0,2.3)

Ahora, queremos *definir* operaciones que involucren varibles tipo `Vec2D`, en particular la suma:

In [65]:
Vec2D(1,2) + Vec2D(0,3)

LoadError: LoadError: MethodError: `+` has no method matching +(::Vec2D{Int64}, ::Vec2D{Int64})
Closest candidates are:
  +(::Any, ::Any, !Matched::Any, !Matched::Any...)
while loading In[65], in expression starting on line 1

In [66]:
# Lo primero sirve para *no* sobreescribir el operador `+` arbitrariamente
import Base.+

# ... y aquí se implementa la nueva función
+(a::Vec2D, b::Vec2D) = Vec2D( a.x + b.x, a.y + b.y )

+ (generic function with 150 methods)

**NOTA** No es necesario que la parte derecha incluya el mismo operador; puede corresponder a cualquier locura que se les ocurra...

In [67]:
Vec2D(1,2) + Vec2D(0,3)

Vec2D{Int64}(1,5)

In [68]:
Vec2D(1,2.2) + Vec2D(-1,3)

Vec2D{Float64}(0.0,5.2)

In [70]:
+3

3

In [72]:
+(a::Vec2D) = a

+ (generic function with 151 methods)

In [73]:
+ Vec2D(-1,3)

Vec2D{Int64}(-1,3)

De igual manera se pueden definir los otros operadores, o incluso inventar nuevos. Un ejemplo es el siguiente:

In [74]:
..(a::Real, b::Real) = Vec2D(a, b)

.. (generic function with 1 method)

In [75]:
..(1.2,3)

Vec2D{Float64}(1.2,3.0)

Y, lo anterior, en ciertos casos se puede escribir de manera más bonita:

In [76]:
1.2 .. 3

Vec2D{Float64}(1.2,3.0)

In [77]:
methods(..)

# 1 method for generic function "..":
..(a::Real, b::Real) at In[74]:1

In [78]:
import Base.*

*(a::Vec2D, b::Vec2D) = Vec2D(a.x, b.y)

* (generic function with 135 methods)

In [80]:
aa = Vec2D(1, 2.4)
bb = Vec2D(-6, 5//1)
println(aa,bb)

Vec2D{Float64}(1.0,2.4)Vec2D{Rational{Int64}}(-6//1,5//1)


In [81]:
aa * bb

Vec2D{Float64}(1.0,5.0)

In [82]:
bb * aa

Vec2D{Float64}(-6.0,2.4)

En particular, uno puede construir funciones genéricas que, si uno *sobrecarga* los operadores de manera adecuada, puede extender por ejemplo a `Vec2D`.

In [83]:
f(x, y) = 2x + y

f (generic function with 1 method)

In [84]:
f(2,3)

7

In [85]:
f(aa, bb)

LoadError: LoadError: MethodError: `*` has no method matching *(::Int64, ::Vec2D{Float64})
Closest candidates are:
  *(::Any, ::Any, !Matched::Any, !Matched::Any...)
  *(::Real, !Matched::Complex{T<:Real})
  *{N}(::Integer, !Matched::Base.IteratorsMD.CartesianIndex{N})
  ...
while loading In[85], in expression starting on line 1

In [ ]:
#Vec2D(a) = Vec2D(a, -a)

In [88]:
*(a::Real, b::Vec2D) = Vec2D(a*b.x, a*b.y)
*(b::Vec2D, a::Real) = a*b

* (generic function with 137 methods)

In [89]:
f(aa, bb)

Vec2D{Float64}(-4.0,9.8)

In [92]:
"perro" + "gato"

LoadError: LoadError: MethodError: `+` has no method matching +(::ASCIIString, ::ASCIIString)
Closest candidates are:
  +(::Any, ::Any, !Matched::Any, !Matched::Any...)
while loading In[92], in expression starting on line 1